File: 2025/07/main.py
import time


def main():
    grid = get_grid()
    t = time.time()
    sol = count_paths(grid, starting_point(grid), {})
    t2 = time.time()

    print(f"Time: {t2 - t:.6f} seconds")  # Python is 5 times slower than Go here
    print(sol)


def count_paths(grid, point: tuple[int, int], cache: dict) -> int:
    if point in cache:
        return cache[point]

    if grid[point[1]][point[0]] == "^":
        left = (point[0] - 1, point[1])
        left_v = count_paths(grid, left, cache)
        cache[left] = left_v

        right = (point[0] + 1, point[1])
        right_v = count_paths(grid, right, cache)
        cache[right] = right_v

        total = left_v + right_v
        cache[point] = total
        return total
    else:
        if len(grid) <= point[1] + 1:
            cache[point] = 1
            return 1
        bottom = (point[0], point[1] + 1)
        return count_paths(grid, bottom, cache)


def get_grid() -> list[list[str]]:
    grid = []
    with open("in.txt") as file:
        for line in file:
            grid.append([x for x in line.strip()])
    return grid


def starting_point(grid: list[list[str]]) -> tuple[int, int]:
    for y in range(len(grid)):
        for x in range(len(grid[y])):
            if grid[y][x] == "S":
                return (x, y)
    raise ValueError("No starting point found")


if __name__ == "__main__":
    main()

File: 2025/07/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
)

func main() {
	file, _ := os.ReadFile("in.txt")
	grid := NewGrid(file)
	sol := sol2(grid)
	// grid.Show() // can be used in sol1
	fmt.Println(sol)
}

func (grid Grid) countPaths(point *Point, cache Cache) int {
	v, ok := cache[*point]
	if ok {
		return v
	}

	switch grid[point.y][point.x] {
	case '^':
		var leftCount, rightCount int
		leftPoint, err := point.GetLeftPoint(grid)
		if err == nil {
			leftCount = grid.countPaths(leftPoint, cache)
			cache[*leftPoint] = leftCount
		}
		rightPoint, err := point.GetRightPoint(grid)
		if err == nil {
			rightCount = grid.countPaths(rightPoint, cache)
			cache[*rightPoint] = rightCount
		}
		sum := leftCount + rightCount
		cache[*point] = sum
		return sum
	default:
		nextPoint, err := point.GetBottomPoint(grid)
		if err != nil {
			cache[*point] = 1
			return 1
		}
		return grid.countPaths(nextPoint, cache)
	}
}

func sol2(grid Grid) int {
	cache := Cache{}
	return grid.countPaths(grid.StartingPoint(), cache)
}

func sol1(grid Grid) int {
	sol := 0

	points := map[*Point]struct{}{grid.StartingPoint(): {}}

	for len(points) != 0 {
		nextPoints := map[*Point]struct{}{}

		for point := range points {
			switch grid[point.y][point.x] {
			case 'S':
				nextPoint, _ := point.GetBottomPoint(grid)
				nextPoints[nextPoint] = struct{}{}
			case '.':
				grid[point.y][point.x] = '|'
				nextPoint, err := point.GetBottomPoint(grid)
				if err != nil {
					continue
				}
				nextPoints[nextPoint] = struct{}{}
			case '^':
				sol++
				leftPoint, err := point.GetLeftPoint(grid)
				if err == nil {
					nextPoints[leftPoint] = struct{}{}
				}
				rightPoint, err := point.GetRightPoint(grid)
				if err == nil {
					nextPoints[rightPoint] = struct{}{}
				}
			}
		}

		points = nextPoints
	}

	return sol
}

type Grid [][]byte

func (g Grid) Show() {
	for _, row := range g {
		fmt.Println(string(row))
	}
}

func NewGrid(file []byte) Grid {
	lines := strings.Split(string(file), "\n")
	grid := make([][]byte, len(lines))

	for rowIndex, line := range lines {
		row := []byte(line)
		grid[rowIndex] = row
	}

	return grid
}

func (g Grid) StartingPoint() *Point {
	for x, ch := range g[0] {
		if string(ch) == "S" {
			return &Point{x, 0}
		}
	}
	return nil
}

type Point struct {
	x, y int
}

func (p *Point) GetBottomPoint(grid Grid) (*Point, error) {
	if p.y+1 >= len(grid) {
		return nil, PointOutsideGridError
	}
	return &Point{p.x, p.y + 1}, nil
}

func (p *Point) GetRightPoint(grid Grid) (*Point, error) {
	if p.x+1 >= len(grid[0]) {
		return nil, PointOutsideGridError
	}
	return &Point{p.x + 1, p.y}, nil
}

func (p *Point) GetLeftPoint(grid Grid) (*Point, error) {
	if p.x <= 0 {
		return nil, PointOutsideGridError
	}
	return &Point{p.x - 1, p.y}, nil
}

var PointOutsideGridError = errors.New("Point goes outside of grid")

type Cache map[Point]int
